pandas/core/reshape/merge.py

```diff
@@ -962,14 +963,30 @@ def _maybe_coerce_merge_keys(self):
                 elif lib.infer_dtype(lk) == lib.infer_dtype(rk):
                     pass
 
-            # Houston, we have a problem!
-            # let's coerce to object if the dtypes aren't
-            # categorical, otherwise coerce to the category
-            # dtype. If we coerced categories to object,
-            # then we would lose type information on some
-            # columns, and end up trying to merge
-            # incompatible dtypes. See GH 16900.
             else:
+
+                # Check if we are trying to merge on obviously
+                # incompatible dtypes GH 9780
+                msg = ("You are trying to merge on {lk_dtype} and "
```

Comment:
can you do these as separate elif (rather than catch these on the big else, which should be just for the object type)